File: lab2/mandelbrot.py
from OpenGL.GL import *

import utils
import math

dimensions = utils.Coord(400, 400)
bounds = utils.Coord(dimensions.x / 2, dimensions.y / 2)
viewport = utils.Viewport(bounds)

background = [0.5, 0.5, 0.5]


def startup():
    glClearColor(background[0], background[1], background[2], 0)
    viewport(None, dimensions)
    make_mandelbrot_set(-2.0, 1.0, 1.5, -1.5)


def shutdown():
    pass


def is_convergent(point: utils.Coord, max_steps=50):
    z = complex(0, 0)
    p = complex(point.x, point.y)
    for _ in range(max_steps):
        z = z ** 2 + p
        if math.sqrt(z.real ** 2 + z.imag ** 2) >= 2:
            return False
    return True


def make_mandelbrot_set(left, right, top, bottom):
    glColor3ub(255, 0, 0)
    glBegin(GL_POINTS)
    for x in range(int(dimensions.x)):
        for y in range(int(dimensions.y)):
            coord = utils.Coord(left + (x * (right - left) / dimensions.x),
                                top + (y * (bottom - top) / dimensions.y))
            if is_convergent(coord):
                glVertex2i(int(x - bounds.x), int(y - bounds.y))

    glEnd()


def main():
    utils.one_frame_init(dimensions=dimensions,
                         startup=startup,
                         shutdown=shutdown,
                         viewport=viewport)


if __name__ == '__main__':
    main()

File: lab2/Mandelbrot.cpp
#include <GLFW/glfw3.h>
#include <complex>

#include "Mandelbrot.h."

namespace gk::lab3
{

auto Mandelbrot::run(Vector2i dimensions, bool vsync) -> void
{
    this->dimensions = dimensions;
    this->bounds = Vector2f {static_cast<float>(dimensions.x) / 2.f, static_cast<float>(dimensions.y) / 2.f};

    utils::configureViewport(bounds, nearDepth, farDepth);
    utils::defaultInit(dimensions,
                       [this] { startup(); },
                       [this](auto time) { render(time); },
                       [this] { shutdown(); },
                       utils::Dimensions::_2D,
                       vsync);
}

auto Mandelbrot::startup() const -> void
{
    glClearColor(0.5f, 0.5f, 0.5f, 0.1f);
    utils::viewport(nullptr, dimensions.x, dimensions.y);
}

auto Mandelbrot::shutdown() -> void
{

}

auto Mandelbrot::render([[maybe_unused]] double time) -> void
{
    glClear(GL_COLOR_BUFFER_BIT);
    makeMandelbrotSet({-2.f, 1.5f}, {1.f, -1.5f});
    glFlush();
}

auto Mandelbrot::makeMandelbrotSet(Vector2f topLeft, Vector2f bottomRight) const -> void
{
    glColor3ub(56, 127, 230);
    glBegin(GL_POINTS);
    for (auto i = 0; i < dimensions.x; i++)
    {
        for (auto j = 0; j < dimensions.y; j++)
        {
            const auto coord = Vector2f {topLeft.x + (static_cast<float>(i) * (bottomRight.x - topLeft.x) / static_cast<float>(dimensions.x)),
                                         topLeft.y + (static_cast<float>(j) * (bottomRight.y - topLeft.y) / static_cast<float>(dimensions.y))};

            if (isConvergent(coord, 30))
            {
                glVertex2i(i - static_cast<int>(bounds.x), j - static_cast<int>(bounds.y));
            }
        }
    }
    glEnd();
}

auto Mandelbrot::isConvergent(Vector2f point, uint32_t maxSteps) -> bool
{
    auto z = std::complex<float>{};
    const auto p = std::complex{point.x, point.y};
    for (auto i = uint32_t{}; i < maxSteps; i++)
    {
        z = std::pow(z, 2) + p;

        if (std::abs(z) >= 2.f)
        {
            return false;
        }
    }
    return true;
}

}